Add the database migration service

tczekajlo opened the pull request:
Closes #71

wochinge commented:
@tczekajlo Sorry, I'm a bit swamped at the moment. Is it urgent? Trying to have a look tomorrow or Monday 😬

tczekajlo commented:
> @tczekajlo Sorry, I'm a bit swamped at the moment. Is it urgent? Trying to have a look tomorrow or Monday 😬

No rush, we support the database migration service for >= 0.33.0 version :)

wochinge commented on charts/rasa-x/templates/db-migration-service.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: "v1"
+kind: "Service"
+metadata:
+  name: {{ include "rasa-x.fullname" . }}-db-migration-service-headless
```
should we make the service name configurable?

tczekajlo replied:
Hmm, It's hard for me to imagine a case that we'd need to change the service name, I'd leave it as it is 😄

wochinge commented on charts/rasa-x/templates/db-migration-service.yaml:
```diff
@@ -0,0 +1,21 @@
+    name: "http"
+  selector:
+    {{- include "rasa-x.selectorLabels" . | nindent 4 }}
+    app.kubernetes.io/component: db-migration-service
```
haha, the `service` confused me very much here 😄

tczekajlo replied:
That's true 😄

wochinge commented on charts/rasa-x/templates/db-migration-service-deployment.yaml:
```diff
@@ -0,0 +1,77 @@
+        - name: "SELF_PORT"
+          value: {{ .Values.dbMigrationService.port | quote }}
+        - name: "RUN_DATABASE_MIGRATION_AS_SEPARATE_SERVICE"
+          value: {{ .Values.separateDBMigrationService | quote }}
```
should be always this, shouldn't it?
```suggestion
          value: "true"
```

tczekajlo replied:
Good point 👍

wochinge commented on charts/rasa-x/templates/rasa-x-deployment.yaml:
```diff
@@ -34,6 +34,7 @@ spec:
       tolerations:
 {{ toYaml .Values.rasax.tolerations | indent 6 }}
       {{- end }}
+      {{- include "initContainer.dbMigration" . |  nindent 6 }}
```
Do we even need this here? The container won't be killed as the status endpoint should return a correct status, no?

tczekajlo replied:
You're right, we don't need it here

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -157,6 +157,53 @@ rasa:
+  name: "rasa/rasa-x"  # gcr.io/rasa-platform/rasa-x-ee
+  # tag refers to the Rasa X image tag (uses `appVersion` by default)
+  tag: ""
+  # don't check required minimum version and run the service
```
?
```suggestion
  # don't check required minimum Rasa X version and run the service
```

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -157,6 +157,53 @@ rasa:
       # resources which rasaWorker is required / allowed to use
       resources: {}
 
+# The database migration service required Rasa X >= 0.33.0
```
```suggestion
# The database migration service requires Rasa X >= 0.33.0
```

wochinge commented on charts/rasa-x/templates/rasa-deployments.yaml:
```diff
@@ -31,6 +31,7 @@ spec:
       tolerations:
 {{ toYaml $.Values.rasa.tolerations | indent 6 }}
       {{- end }}
+      {{- include "initContainer.dbMigration" $ |  nindent 6 }}
```
why?

tczekajlo replied:
I use the init container for the rasa deployment because the liveness probe for the rasa-production and rasa-worker service doesn't pass as long as the rasa-x is not fully operational.  

To make sure that the services are not restarting and have a clear start, first, we check for a DB migration and we make sure that everything is ready to go!

Without the init container:
```
rasa-x-rasa-production-65c964c9df-6h88r   1/1     Running   1          2m48s
rasa-x-rasa-worker-5d766bf7c4-5tpkr       1/1     Running   1          2m48s
```

with the init container:
```
rasa-x-rasa-production-678b9bb5f8-dntpm        1/1     Running            0          18m
rasa-x-rasa-worker-86b6498c6c-nqsxt            1/1     Running            0          18m
```

wochinge replied:
Makes sense, but it's stupid, isn't it? I think it would be better if Rasa Open Source starts independent of what Rasa X. does and pulls the config later. Wdyt? 

We should also add your comment as comment to the deployment file to make it clear why we need the init container there.

tczekajlo replied:
> Makes sense, but it's stupid, isn't it? I think it would be better if Rasa Open Source starts independent of what Rasa X. does and pulls the config later. Wdyt?

I see your point and I agree with you that it'd be better to start Rasa OS independently.

From the perspective of a user who uses the rasa-x helm chart first time, currently, after deploy, I see two services that ware restarted. The first reaction will be to check the logs of the services, but the logs are empty, zero information on why the services ware restarted.

I think if we can save someone from scratching in the head and wondering what happened, by simply adding the init container we should that.

wochinge commented on charts/rasa-x/templates/_helpers.tpl:
```diff
@@ -206,3 +215,46 @@ Return the AppVersion value as a default if the app.tag variable is not defined.
+{{/*
+Return 'true' if required version to run the database migration service is correct.
+*/}}
+{{- define "db-migration-service.requiredVersion" -}}
```
very neat idea with the version check!

tczekajlo replied:
Thanks 😄

wochinge reviewed (commented):
nice job and some great ideas (e.g. the one about validating the Rasa X version) 🚀 
Commenting for now as I have some questions and we need to wait for the 0.33 release

tczekajlo commented on charts/rasa-x/templates/db-migration-service-deployment.yaml:
```diff
@@ -0,0 +1,77 @@
+        {{- include "rasa-x.selectorLabels" . | nindent 8 }}
+        app.kubernetes.io/component: db-migration-service
+    spec:
+      automountServiceAccountToken: false
```
It's a good security practice to not mount a service account token if it's not used.

wochinge replied:
should we create an issue / separate PR to also do this for the other deployments?

tczekajlo commented:
The Rasa X 0.33.0 is released, so it's time to go back to the PR 😄  I've added two things that came up in the meantime:
- Semantic version - if the `rasax.tag` is not a valid semantic version then not use the DB migration service. It's still possible to use it by means of the `dbMigrationService.ignoreVersionCheck=true`
- Wait for DB migration - some tests, scripts (RasaHQ/rasa-x#3902) don't check if DB migration is completed and try to send a request too early. I've added the init container to the rasa-x deployment in order to wait for the DB migration to be done. It's possible to disable it by using the `useDBinitContainer=false`.

@wochinge Can you 👀  if you have a minute? 🙌

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -194,6 +199,53 @@ rasa:
+# The database migration service requires Rasa X >= 0.33.0
+# db-migration-service specific settings
+dbMigrationService:
+  # settings related to the init-db container used as a init container for services
```
Should we follow the recommended Helm best practices and repeat the property name at the start of each comment? https://helm.sh/docs/chart_best_practices/values/#document-valuesyaml

tczekajlo replied:
Sure good point 👍

wochinge reviewed (commented):
A few tiny things - looks great overall 💪 

> Wait for DB migration - some tests, scripts (RasaHQ/rasa-x#3902) don't check if DB migration is completed and try to send a request too early. I've added the init container to the rasa-x deployment in order to wait for the DB migration to be done. It's possible to disable it by using the rasax.useDBinitContainer=false.

One reason for the introduction of the separate migration service was that users can already access the Rasa X UI which can inform them that the deployment worked so far, but the database migrations aren't finished yet. When we add the init container to the Rasa X deployment then users won't know what's going on any longer. Hence, I'd prefer it we would adapt our scripts instead of adding the init container. What do you think?

tczekajlo commented:
> A few tiny things - looks great overall 💪
> 
> > Wait for DB migration - some tests, scripts ([RasaHQ/rasa-x#3902](https://github.com/RasaHQ/rasa-x/issues/3902)) don't check if DB migration is completed and try to send a request too early. I've added the init container to the rasa-x deployment in order to wait for the DB migration to be done. It's possible to disable it by using the rasax.useDBinitContainer=false.
> 
> One reason for the introduction of the separate migration service was that users can already access the Rasa X UI which can inform them that the deployment worked so far, but the database migrations aren't finished yet. When we add the init container to the Rasa X deployment then users won't know what's going on any longer. Hence, I'd prefer it we would adapt our scripts instead of adding the init container. What do you think?

You're right, I was thinking about it a bit more and I agree with you.

tczekajlo commented on charts/rasa-x/templates/db-migration-service-deployment.yaml:
```diff
@@ -0,0 +1,77 @@
+{{ if and (eq "true" (include "db-migration-service.requiredVersion" .)) .Values.separateDBMigrationService }}
+apiVersion: apps/v1
+kind: Deployment
```
Hmm, actually it's a good point 👍

tczekajlo commented:
@wochinge Thanks a lot for reviewing 🙌   I've made changes based on the code review. Can you give another look if you have a minute? 🙇

wochinge commented on .github/workflows/ci.yml:
```diff
@@ -97,4 +104,8 @@ jobs:
       run: |
-        curl --fail --data '{"username": "me", "password": "'${INITIAL_USER_PASSWORD}'"}' "http://localhost/api/auth"
+        TOKEN=$(curl --fail --data '{"username": "me", "password": "'${INITIAL_USER_PASSWORD}'"}' "http://localhost/api/auth" | jq -r '.access_token')
+        if [[ "${TOKEN}" == "" ]]; then
```
?
```suggestion
        if [[ -z "${TOKEN}" ]]; then
```

wochinge commented on charts/rasa-x/templates/_helpers.tpl:
```diff
@@ -210,3 +219,53 @@ Return 'true' if an enterprise version is run.
+  - "apk update --no-cache && \
+    apk add jq curl && \
+    until nslookup {{ .Release.Name }}-db-migration-service-headless 1> /dev/null; do echo Waiting for the database migration service; sleep 2; done && \
+    until [[ \"$(curl -s http://{{ .Release.Name }}-db-migration-service-headless:{{ .Values.dbMigrationService.port }} | jq -r .status)\" == \"completed\" ]]; do \
```
should we use a template for `http://{{ .Release.Name }}-db-migration-service-headless:{{ .Values.dbMigrationService.port }}` as we are repeating this multiple times?

wochinge commented on charts/rasa-x/templates/db-migration-service.yaml:
```diff
@@ -0,0 +1,21 @@
+{{ if eq "true" (include "db-migration-service.requiredVersion" .) }}
+{{ if .Values.separateDBMigrationService }}
```
optional: slightly inconsistent to the other uses of `and` instead of two `if`s for this condition

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -194,6 +194,54 @@ rasa:
+# dbMigrationService specifies settings for the database migration service
+# The database migration service requires Rasa X >= 0.33.0
+dbMigrationService:
+  # settings related to the init-db container used as a init container for services
```
```suggestion
  # initContainer describes settings related to the init-db container used as a init container for deployments
```

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -194,6 +194,54 @@ rasa:
+dbMigrationService:
+  # settings related to the init-db container used as a init container for services
+  initContainer:
+    image: alpine:latest
```
```suggestion
    # image is the Docker image which is used by the init container
    image: alpine:latest
```

wochinge commented on charts/rasa-x/values.yaml:
```diff
@@ -194,6 +194,54 @@ rasa:
+  command: []
+  # args overrides the default arguments to run in the container
+  args: []
+  # name is a Docker image name
```
optional: Just to make sure users are not confused between the init container image and this one
```suggestion
  # name is the Docker image name which is used by the migration service
```

wochinge commented on charts/rasa-x/templates/_helpers.tpl:
```diff
@@ -210,3 +219,53 @@ Return 'true' if an enterprise version is run.
+If version is not valid semantic version then not use the DB migration service.
+*/}}
+{{- define "db-migration-service.requiredVersion" -}}
+{{- if and (not (regexMatch "^(v[0-9]|0|[1-9]\\d*).(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$" (include "db-migration-service.version" .))) (not .Values.dbMigrationService.ignoreVersionCheck) -}}
```
> ^(v[0-9]|0|[1-9]

Isn't this enough `^(v[0-9]+.` ? (similar for some of the following).

 And do we actually need such a check? `semverCompare` will fail anyway if it's not a valid version, no?

tczekajlo replied:
Actually, the regex is copy & paste from the https://semver.org/ website, so I assume it's the best one to check semantic version 😄   The check purpose is to deal with a scenario where someone uses a custom non-sematic tag for a Rasa X image, instead of throwing an error that a version is not valid, we assume that the version is not the required version to run the DB migration service.  Wdyt?

wochinge replied:
Makes sense! Wouldn't it be better to fail loud and early instead of silently disabling it?

tczekajlo replied:
I was a bit driven by the https://github.com/RasaHQ/rasa-x-helm/issues/86 issue, but you're right. It's better to fail loudly. Additionally, there is the `dbMigrationService.ignoreVersionCheck` parameter that can be used to disable the check. Thanks 🙌

wochinge replied:
I'll do a PR for the #86 later this week 👍

wochinge reviewed (approved):
Great and clean change! 🚀
